hack/reduce_nat_gateway_cost/replace_nat_with_nat_instance.py

```diff
@@ -0,0 +1,1250 @@
+# 2. A cluster may run in a single AZ or split itself across multiple AZs for fault tolerance.
+#    For example, it may put one master node in us-east-1a, another in us-east-1b, and another
+#    in us-east-1c. For each AZ for which a cluster is configured to span. the installation process will
+#    create:
+#    a. Two subnets. One public and one private. A public subnet is defined as one that possesses
+#       an Internet Gateway (IGW). An IGW allows traffic to flow between EC2 instances attached to
+#       a public subnet and the Internet IFF the EC2 instance has a public IP address assigned.
+#       A private subnet does not have an IGW. EC2 Instances attached to a private subnet, even
+#       if they had a public IP address cannot not communicate directly to the Internet with it.
+#       Instead, these instances have private IP addresses (e.g. 10.0.29.172) and use a NAT
+#       gateway to perform addresses translate to communicate on the Internet. If you curl a "what is my IP"
+#       service from an OpenShift node, you will see the public IP address of the NAT gateway,
+#       in the AZ for the node and not the node itself (which does not have a public ipv4 in a standard installation).
+#    b. A NAT gateway attached to the private subnet.
+#    c. An IGW attached to the public network.
+#    d. Zero or more master or worker nodes attached to the private network.
+#    e. Two RouteTables. One for the public subnet and one for the private. The RouteTable
+#       tells AWS how to route network traffic from an instance attached to a subnet. The
+#       most important routes in this solution are those in the private subnet's RouteTable.
+#       An entry that routes traffic within the private subnet: e.g. 10.0.0.0/19 => local.
+#       And an entry that routes outbound traffic to the Internet: 0.0.0.0/0 => NAT Gateway instance Id.
+#       This entry is used to route traffic for everything NOT in the private subnet's CIDR.
+#
+#  The goal of this lambda is to replace this 0.0.0.0/0 route in the private subnet with
+#  0.0.0.0/0 => EC2 NAT instance network interface (ENI) where the EC2 instance has been enabled
+#  to provide a NAT services.
+#
+# With the knowledge that ephemeral IPI clusters are being created in the AWS account, the
+# lambda takes action to provision a NAT instance in each public subnet being used by a cluster.
+# Once the NAT instance is ready to handle traffic, it will update the RouteTable for
+# the corresponding private subnet associated with the cluster & AZ to route 0.0.0.0/0 traffic
+# (i.e. traffic destined for the Internet) to itself. The NAT instance resides in the
+# public network and uses the IGW to communicate to the internet on behalf of other
+# instances in the cluster to the internet.
+# If the instance is shutdown gracefully, it will reverse this process (by way of a systemd unit
+# installed by userData on startup) and restore the former route for the NAT gateway to handle
+# the traffic.
```

Comment:
if it's not shut down gracefully (reaped by AWS, accidentally deleted, etc) is there a watchdog to restore the former route?

Reply:
to answer my own question... the NAT instance is dedicated to a cluster AZ so only affects that cluster if something goes wrong. very low blast radius.

Reply:
Adding:

> .... As a backup for a graceful shutdown, the lambda itself will also listen for
> TerminateInstances API calls and, if a master or the NAT instance is targeted, the lambda
> will restore the NAT Gateway route.